### PR TITLE
[GLUTEN-3559][VL] Fixed unit test failing after change of Spark-38674

### PR DESCRIPTION
**Pull request by vibhaska**
## What changes were proposed in this pull request?

This PR fixes the following suites:
GlutenDynamicPartitionPruningV1SuiteAEOff
GlutenDynamicPartitionPruningV1SuiteAEOff
GlutenDynamicPartitionPruningV1SuiteAEOn
GlutenDynamicPartitionPruningV1SuiteAEOnDisableScan
GlutenDynamicPartitionPruningV1SuiteAEOffDisableScan
GlutenDynamicPartitionPruningV2SuiteAEOff
GlutenDynamicPartitionPruningV2SuiteAEOn
GlutenDynamicPartitionPruningV2SuiteAEOnDisableScan
GlutenDynamicPartitionPruningV2SuiteAEOffDisableScan

(Fixes: \#3559)

## How was this patch tested?

Ran these UTs to confirm the fix

### CODE REVIEW

**Comment by vibhaska** [comment]
@JkSelf Kindly review.
Some failing checks seems unrelated.

**Comment by ayushi-agarwal** [comment]
@zhli1142015 @ulysses-you Could you please help in reviewing this?

File: gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala [hunk 1]
```diff
@@ -108,6 +108,7 @@ case class ColumnarSubqueryBroadcastExec(
         }
         val dataSize = rows.map(_.asInstanceOf[UnsafeRow].getSizeInBytes).sum
         longMetric("dataSize") += dataSize
+        metrics("numOutputRows").add(rows.length)
```
Comment (zhli1142015) [inline gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala:111]: nit, please follow line 110.

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala [hunk 2]
```diff
@@ -550,17 +606,26 @@ class GlutenDynamicPartitionPruningV1SuiteAEOff
           scanOption.get
         }
 
+        def getDriverMetrics(plan: SparkPlan, key: String): Option[SQLMetric] = {
```
Comment (zhli1142015) [inline gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala:609]: Looks this function is duplicated several times. If any reason we need it as we always pass scan for `plan`.

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala [hunk 3]
```diff
@@ -451,17 +484,26 @@ abstract class GlutenDynamicPartitionPruningV1Suite extends GlutenDynamicPartiti
...
         val df1 = sql("SELECT sum(f1) FROM fact")
         df1.collect()
         val scan1 = getFactScan(df1.queryExecution.executedPlan)
         assert(!scan1.metrics.contains("staticFilesNum"))
         assert(!scan1.metrics.contains("staticFilesSize"))
-        val allFilesNum = scan1.metrics("numFiles").value
-        val allFilesSize = scan1.metrics("filesSize").value
-        assert(scan1.metrics("numPartitions").value === numPartitions)
-        assert(scan1.metrics("pruningTime").value === -1)
+        val allFilesNum: Long = getDriverMetrics(scan1, "numFiles").map(_.value).getOrElse(-1)
```
Comment (JkSelf) [inline gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala:499]: @vibhaska It seems the scan1 is the `FileSourceScanExecTransformer`. Why need to find the `FileSourceScanExe `plan metrics here ?

**Review by JkSelf — COMMENTED** [review]
@vibhaska Thanks for your great work. Just two small questions. Thanks.

File: gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala (hunk 1 above)
Comment (vibhaska) [reply gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarSubqueryBroadcastExec.scala]: DONE

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala (hunk 2 above)
Comment (vibhaska) [reply gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala]: @zhli1142015 he corresponding metrics need to e sourced from driverMetrics (as against metrics, used earlier)

File: gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala (hunk 3 above)
Comment (vibhaska) [reply gluten-ut/spark34/src/test/scala/org/apache/spark/sql/GlutenDynamicPartitionPruningSuite.scala:500]: @JkSelf Updated implementation of getDriverMetrics. Idea is to getdriverMetrics from FileSourceScanExe .

**Review by JkSelf — APPROVED** [review]
LGTM. Can you help to update the issue list [here](https://github.com/oap-project/gluten/issues/3559#issue-1967407941)?

**Pull request merged**